library/ceph_crush.py

```diff
@@ -0,0 +1,211 @@
+        description:
+            - osd_crush_location dict from the inventory file. It contains
+            the placement of each host in the CRUSH map.
+        required: true
+    containerized:
+        description:
+            - Weither or not this is a containerized cluster. The value is
+            assigned or not depending on how the playbook runs.
+        required: false
+        default: None
+'''
+
+EXAMPLES = '''
+- name: configure crush hierarchy
+  ceph_crush:
+    cluster: "{{ cluster }}"
+    location: "{{ hostvars[item]['osd_crush_location'] }}"
+    containerized: "{{ docker_exec_cmd }}"
+  with_items: "{{ groups[osd_group_name] }}"
+  when:
+    - crush_rule_config
+'''
+
+RETURN = '''#  '''
+
+from ansible.module_utils.basic import AnsibleModule
+import datetime
+
+
+def fatal(message, module):
+    '''
+    Report a fatal error and exit
+    '''
+    if module:
+        module.fail_json(msg=message, rc=1)
+    else:
+        raise(Exception(message))
+
+
+def generate_cmd(cluster, subcommand, bucket, bucket_type, containerized=None):
+    '''
+    Generate command line to execute
+    '''
+    cmd = [
+        'ceph',
+        '--cluster',
+        cluster,
+        'osd',
+        'crush',
+        subcommand,
+        bucket,
+        bucket_type,
+    ]
+    if containerized:
+        for piece in reversed(containerized.split()):
+            cmd.insert(0, piece)
+    return cmd
+
+
+def sort_osd_crush_location(location, module):
+    '''
+    Sort location tuple
+    '''
+    if len(location) < 2:
+        fatal("You must specify at least 2 buckets.", module)
+    [item for item in location if item[0] == "host"] or fatal("You must specify a 'host' bucket.", module)
+
+    try:
+        crush_bucket_types = [
+            "host",
+            "chassis",
+            "rack",
+            "row",
+            "pdu",
+            "pod",
+            "room",
+            "datacenter",
+            "region",
+            "root",
+        ]
+        return sorted(location, key=lambda crush: crush_bucket_types.index(crush[0]))
+    except ValueError as error:
+        fatal("{} is not a valid CRUSH bucket, valid bucket types are {}".format(error.args[0].split()[0], crush_bucket_types), module)
+
+
+def create_and_move_buckets_list(cluster, location, containerized=None):
+    '''
+    Creates Ceph CRUSH buckets and arrange the hierarchy
+    '''
+    previous_bucket = None
+    cmd_list = []
+    for item in location:
+        bucket_type, bucket_name = item
+        # ceph osd crush add-bucket maroot root
+        cmd_list.append(generate_cmd(cluster, "add-bucket", bucket_name, bucket_type, containerized))
+        if previous_bucket:
+            # ceph osd crush move monrack root=maroot
+            cmd_list.append(generate_cmd(cluster, "move", previous_bucket, "%s=%s" % (bucket_type, bucket_name), containerized))
+        previous_bucket = item[1]
+    return cmd_list
+
+
+def exec_commands(module, cmd_list):
```

Comment:
I'm a bit confused by this method. So this is taking a list of commands, running them all and only returning the ``rc``,``out`` and ``err`` of the last command run? Is that the expected behavior?

Reply:
Yes, I know this is not ideal as we don't have the full rc/out/err for all the commands that were executed. However, if something goes wrong we will catch it, if everything is fine then we will get the last command.